<table class="tfo-notebook-buttons" align="left">
<td style="border: none;">
<a target="blank" href="https://colab.research.google.com/github/Hari31416/Portfolio/blob/main/ML\Disaster_Tweets\Preprocessing.ipynb?hl=en"><img src="https://colab.research.google.com/img/colab_favicon_256px.png" width="28" height="28"/>Run on Google Colab</a></td>
<td style="border: none;">
<a target="blank" href="https://github.com/Hari31416/Portfolio/blob/main/ML\Disaster_Tweets\Preprocessing.ipynb"><img src="https://cdn.icon-icons.com/icons2/2368/PNG/512/github_logo_icon_143772.png" width="28" height="28"/>View on Github</a></td>
<td style="border: none;">
</table>


<h1 id="Contents">Contents<a href="#Contents"></a></h1>
        <ol>
        <li><a class="" href="#Imports">Imports</a></li>
<li><a class="" href="#Getting-Data-Ready">Getting Data Ready</a></li>
<ol><li><a class="" href="#Downloading-the-Data">Downloading the Data</a></li>
<li><a class="" href="#Extracting-the-Data">Extracting the Data</a></li>
<li><a class="" href="#Loading-the-Data">Loading the Data</a></li>
</ol><li><a class="" href="#Preprocessing">Preprocessing</a></li>
<ol><li><a class="" href="#Null-Values">Null Values</a></li>
<ol><li><a class="" href="#location-Column">location Column</a></li>
<li><a class="" href="#Keyword-Column">Keyword Column</a></li>
</ol><li><a class="" href="#Cleaning-the-Text-Column">Cleaning the Text Column</a></li>
<ol><li><a class="" href="#Saving-the-Cleaned-Data">Saving the Cleaned Data</a></li>
</ol>

# Imports

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Getting Data Ready

## Loading the Data

In [17]:
train = pd.read_csv('data/train.csv')
test = pd.read_csv('data/test.csv')

In [18]:
train.head()

id keyword location                                               text  \
0   1     NaN      NaN  Our Deeds are the Reason of this #earthquake M...   
1   4     NaN      NaN             Forest fire near La Ronge Sask. Canada   
2   5     NaN      NaN  All residents asked to 'shelter in place' are ...   
3   6     NaN      NaN  13,000 people receive #wildfires evacuation or...   
4   7     NaN      NaN  Just got sent this photo from Ruby #Alaska as ...   

   target  
0       1  
1       1  
2       1  
3       1  
4       1

In [19]:
train.value_counts("target")/len(train)

target
0    0.57034
1    0.42966
dtype: float64

# Preprocessing

## Null Values

In [20]:
train.isna().sum()

id             0
keyword       61
location    2533
text           0
target         0
dtype: int64

In [21]:
test.isna().sum()

id             0
keyword       26
location    1105
text           0
dtype: int64

There are a lot of null values in the data. Let's handle them one by one.

#### `location` Column

In [22]:
train["location"].nunique()

3341

In [23]:
locations = train["location"].unique()
locations[:20]

array([nan, 'Birmingham', 'Est. September 2012 - Bristol', 'AFRICA',
       'Philadelphia, PA', 'London, UK', 'Pretoria', 'World Wide!!',
       'Paranaque City', 'Live On Webcam', 'milky way',
       'GREENSBORO,NORTH CAROLINA', 'England.',
       'Sheffield Township, Ohio', 'India', 'Barbados', 'Anaheim',
       'Abuja', 'USA', 'South Africa'], dtype=object)

There are 3341 unique values in the `location` column. Furthermore, the values of `location` are not very relavant to whether the tweet corresponds to diasaster or not. For now, we will not use this column for further analysis. Let's head to the `keyword` column.

#### `Keyword` Column

In [24]:
train["keyword"].nunique()

221

In [25]:
keywords = train["keyword"].unique()
keywords

array([nan, 'ablaze', 'accident', 'aftershock', 'airplane%20accident',
       'ambulance', 'annihilated', 'annihilation', 'apocalypse',
       'armageddon', 'army', 'arson', 'arsonist', 'attack', 'attacked',
       'avalanche', 'battle', 'bioterror', 'bioterrorism', 'blaze',
       'blazing', 'bleeding', 'blew%20up', 'blight', 'blizzard', 'blood',
       'bloody', 'blown%20up', 'body%20bag', 'body%20bagging',
       'body%20bags', 'bomb', 'bombed', 'bombing', 'bridge%20collapse',
       'buildings%20burning', 'buildings%20on%20fire', 'burned',
       'burning', 'burning%20buildings', 'bush%20fires', 'casualties',
       'casualty', 'catastrophe', 'catastrophic', 'chemical%20emergency',
       'cliff%20fall', 'collapse', 'collapsed', 'collide', 'collided',
       'collision', 'crash', 'crashed', 'crush', 'crushed', 'curfew',
       'cyclone', 'damage', 'danger', 'dead', 'death', 'deaths', 'debris',
       'deluge', 'deluged', 'demolish', 'demolished', 'demolition',
       'derail', 'der

Though the column contain 221 unique values our first thought would be to drop this columns too. However, the keywords are highly relevant to the disaster. Therefore, we will keep this column. How to use this in model, we will see later. First, we need to fill the null values.

Let's see how these keywords are related by the `text`.

In [26]:
keyword_text = train[train["keyword"].notna()][["keyword", "text"]]

In [27]:
import random
indices= keyword_text.index
for i in range(10):
    index = random.choice(indices)
    print(keyword_text.loc[index, "keyword"])
    print(keyword_text.loc[index, "text"])
    print("\n")

casualty
Casualty Team: Ice Cream Recall Sends Chill Through Food Industry http://t.co/6GsAmY6mts


hellfire
day 3 without my phone and due to my slow shite computers it is utter hellfire :'(
Plez send help


displaced
Displaced Persons GN (2014 Image) #1-1ST NM http://t.co/yEJt18sbm0 http://t.co/RcqacN91bE


bridge%20collapse
Leicester_Merc : ICYMI - #Ashes 2015: Australia collapse at Trent Bridge - how Twitter rea‰Û_ http://t.co/HqeWMREysO) http://t.co/y4y8fclJED


emergency%20services
Services are returning to normal #SouthLine after a medical emergency at Yennora and urgent track equipment repairs at Cabramatta earlier.


ambulance
@medic914 @AACE_org I am surprised we still cannot standardised the clinical practice across the 11 NHS ambulance trust.


drought
Thought it was a drought @_ASHJ? http://t.co/V4Br5gjMIY


lightning
@random_tourist it rained. Some guys tree hit by lightning and some jackholes drove onto flooded streets.


mudslide
You've 100% fucked up when Paul says you

Okay, it seems that the `keyword` are taken from the `text`. We can use this fact to fill the null values in the `keyword` column.

In [28]:
keyword_text_null = train[train["keyword"].isna()][["keyword", "text"]]
indices= keyword_text_null.index
for i in range(10):
    index = random.choice(indices)
    print(keyword_text_null.loc[index, "keyword"])
    print(keyword_text_null.loc[index, "text"])
    print("\n")

nan
An IS group suicide bomber detonated an explosives-packed vest in a mosque inside a Saudi special forces headquarters killing 15 people.


nan
Love skiing


nan
Cooool :)


nan
London is cool ;)


nan
Do you like pasta?


nan
Police investigating after an e-bike collided with a car in Little Portugal. E-bike rider suffered serious non-life threatening injuries.


nan
Heat wave warning aa? Ayyo dei. Just when I plan to visit friends after a year.


nan
Heat wave warning aa? Ayyo dei. Just when I plan to visit friends after a year.


nan
The Latest: More Homes Razed by Northern California Wildfire - ABC News http://t.co/YmY4rSkQ3d


nan
this is ridiculous....




To fill the null values, this is what we'll do:
1. Make a list of all the keywords in the `keyword` column.
2. Loop thorough the `text` for which the `keyword` is null.
3. Search if there is a word in the text that is in the list of keywords.
4. If yes, fill the `keyword` column with that word else fill with some other special word.

But before this, let's make some changes to the `keyword` column. We will be removing the final "s" from the keywords.

In [29]:
for word in keywords:
    word = str(word)
    if word.endswith("s"):
        print(word)

body%20bags
burning%20buildings
bush%20fires
casualties
deaths
debris
emergency%20services
eyewitness
fatalities
first%20responders
flames
floods
forest%20fires
hazardous
hostages
injuries
refugees
rescuers
screams
sirens
survivors
weapons
wild%20fires
wounds


In [30]:
def remove_s(word):
    if word is None:
        return None
    word = str(word).lower()
    if word.endswith("s"):
        return word[:-1]
    else:
        return word

train["keyword"] = train["keyword"].apply(remove_s)
test["keyword"] = test["keyword"].apply(remove_s)

In [31]:
train_keywords = train["keyword"].unique()
test_keywords = test["keyword"].unique()
all_keywords = np.concatenate([train_keywords, test_keywords])
all_keywords = np.unique(all_keywords)
len(train_keywords), len(test_keywords), len(all_keywords)

(215, 215, 215)

Looking great!

In [32]:
def fill_keyword(row):
    if row["keyword"] == "nan":
        for keyword in all_keywords:
            if keyword in row["text"].lower():
                return keyword
        return "neutral"
    else:
        return row["keyword"]

In [33]:
train["keyword"] = train[["keyword", "text"]].apply(fill_keyword, axis=1)
test["keyword"] = test[["keyword", "text"]].apply(fill_keyword, axis=1)

In [34]:
train["keyword"].value_counts(), test["keyword"].value_counts()

(weapon                   78
 flood                    76
 death                    74
 body%20bag               74
 siren                    71
                          ..
 neutral                  20
 epicentre                12
 threat                   11
 inundation               10
 radiation%20emergency     9
 Name: keyword, Length: 215, dtype: int64,
 hostage                  32
 siren                    31
 flood                    29
 death                    26
 body%20bag               26
                          ..
 threat                    5
 fatalitie                 5
 radiation%20emergency     5
 inundation                4
 epicentre                 1
 Name: keyword, Length: 215, dtype: int64)

It is working. Now we'll need to clean the `text` column.

## Cleaning the `Text` Column

We can see that we need to remove pucntuations, user tag, external links, and other special characters from the `text` column.

We'll remove the special characters as well as the stopwords.

In [128]:
from nltk.corpus import stopwords
import re
def clean(text):
    #Lowering text
    text = text.lower()

    #Removing the @user
    text = re.sub(r'@[^\s]+', '', text)

    #Removing the links
    text = re.sub(r'http:\S+', '', text)
    text = re.sub(r'https:\S+', '', text)

    #Removing non ASCII characters
    text = re.sub(r'[^\x00-\x7F]+', '', text)
    #Removing punctuation and special characters
    chars = ["!", "#", ",", ".", ":", ";", "?", "-", "~", ")", "(", "=", ">", "<", "/"]
    for char in chars:
        if char in text:
            text = text.replace(char, "")

    stop = stopwords.words('english')
    #To avoid getting empty strings
    if len(text.split()) > 3:
        text = [word for word in text.split() if word not in stop]
    return " ".join(text)

In [129]:
text = """#RockyFire @hari Update  => California Hwy. 20 closed in both directions due to Lake County fire - #CAfire #wildfires"""
clean(text=text)

'rockyfire update california hwy 20 closed directions due lake county fire cafire wildfires'

In [130]:
text = "These boxes are ready to explode! Exploding Kittens finally arrived! gameofkittens #explodingkittens‰Û_ https://t.co/TFGrAyuDC5"
clean(text=text)

'boxes ready explode exploding kittens finally arrived gameofkittens explodingkittens_'

Working good enough!

In [131]:
train['text'] = train['text'].apply(clean)
test['text'] = test['text'].apply(clean)

In [132]:
train[train["text"].str.len() == 0]

id    keyword               location text  target
4497  6394  hurricane  NAWF SIDE POKING OUT             1

In [133]:
test[test["text"].str.len() == 0]

Empty DataFrame
Columns: [id, keyword, location, text]
Index: []

We can see thatone of text entries contain nothing at all. Let's remove this.

In [136]:
train.drop(train[train["text"].str.len() == 0].index, inplace=True)

In [137]:
train[train["text"].str.len() == 0]

Empty DataFrame
Columns: [id, keyword, location, text, target]
Index: []

Great! Now we'll save the data.

### Saving the Cleaned Data

In [138]:
train.to_csv('data/train_clean.csv', index=False)
test.to_csv('data/test_clean.csv', index=False)